src/features2d/descriptor_matcher.rs

```diff
@@ -64,29 +65,33 @@ impl DescriptorMatcherType {
 
 /// Type for matching keypoint descriptors
 #[derive(Debug)]
-pub struct DescriptorMatcher {
+pub struct DescriptorMatcher<'a> {
     value: *mut CDescriptorMatcher,
+    phantom: PhantomData<&'a ()>,
 }
 
-impl Drop for DescriptorMatcher {
+impl<'a> Drop for DescriptorMatcher<'a> {
     fn drop(&mut self) {
         unsafe {
             cv_matcher_drop(self.value);
         }
     }
 }
 
-impl DescriptorMatcher {
+impl<'a> DescriptorMatcher<'a> {
     /// Creates a descriptor matcher of a given type with the default parameters (using default constructor).
-    pub fn new(descriptor_matcher_type: DescriptorMatcherType) -> DescriptorMatcher {
+    pub fn new(descriptor_matcher_type: DescriptorMatcherType) -> Self {
         let descriptor_matcher_type = CString::new(descriptor_matcher_type.as_str()).unwrap();
         let value = unsafe { cv_matcher_new(descriptor_matcher_type.as_ptr()) };
-        DescriptorMatcher { value: value }
+        DescriptorMatcher {
+            value: value,
+            phantom: PhantomData,
+        }
     }
 
     /// Adds descriptors to train a CPU or GPU descriptor collection
-    pub fn add(&mut self, descriptors: &Vec<&Mat>) {
-        let descriptors = descriptors.iter().map(|x| x.inner).collect();
+    pub fn add(&mut self, descriptors: impl IntoIterator<Item = &'a Mat>) {
```

Comment:
why adding `<'a>` on level type? `add<'.a>` doesn't work?

Reply:
I have to do this because none of the matrices can `Drop` before the matcher. The matcher holds references to the matrices, so it would cause a segfault. This tells Rust not to let that happen by polluting the lifetime of the mather with the matrix references. This is how Rust deals with references in a `Vec` as well, except that it isn't taking explicit references in that case.